mono/mini/interp/transform.c

```diff
@@ -4021,29 +4016,25 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			mt = mint_type (ftype);
 
 			/* the vtable of the field might not be initialized at this point */
-			MonoClass *fld_klass = mono_class_from_mono_type_internal (field->type);
+			MonoClass *fld_klass = mono_class_from_mono_type_internal (ftype);
 			mono_class_vtable_checked (domain, fld_klass, error);
 			goto_if_nok (error, exit);
 
-			if (mt == MINT_TYPE_VT) {
-				MonoClass *klass = mono_class_from_mono_type_internal (ftype);
-				int size = mono_class_value_size (klass, NULL);
-				interp_add_ins (td, MINT_STSFLD_VT);
+			if (mono_class_field_is_special_static (field)) {
+				interp_add_ins (td, MINT_STSFLD);
 				td->last_ins->data [0] = get_data_item_index (td, field);
-				POP_VT (td, size);
 			} else {
-				if (mono_class_field_is_special_static (field)) {
-					interp_add_ins (td, MINT_STSFLD);
-					td->last_ins->data [0] = get_data_item_index (td, field);
-				} else {
-					MonoVTable *vtable = mono_class_vtable_checked (domain, field->parent, error);
-					goto_if_nok (error, exit);
-
-					interp_add_ins (td, MINT_STSFLD_I1 + mt - MINT_TYPE_I1);
-					td->last_ins->data [0] = get_data_item_index (td, vtable);
-					td->last_ins->data [1] = get_data_item_index (td, (char*)mono_vtable_get_static_field_data (vtable) + field->offset);
-				}
+				MonoVTable *vtable = mono_class_vtable_checked (domain, field->parent, error);
+				goto_if_nok (error, exit);
+				interp_add_ins (td, (mt == MINT_TYPE_VT) ? MINT_STSFLD_VT : (MINT_STSFLD_I1 + mt - MINT_TYPE_I1));
+				td->last_ins->data [0] = get_data_item_index (td, vtable);
+				td->last_ins->data [1] = get_data_item_index (td, (char*)mono_vtable_get_static_field_data (vtable) + field->offset);
```

Comment:
I was wondering why we wouldn't use `mono_class_static_field_address ()` here: apparently it can trigger a class init, which we don't want here.